sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go

```diff
@@ -2281,13 +2383,24 @@ func (ss *stageState) bundleReady(em *ElementManager, emNow mtime.Time) (mtime.T
 	inputW := ss.input
 	_, upstreamW := ss.UpstreamWatermark()
 	previousInputW := ss.previousInput
-	if inputW == upstreamW && previousInputW == inputW {
-		slog.Debug("bundleReady: unchanged upstream watermark",
-			slog.String("stage", ss.ID),
-			slog.Group("watermark",
-				slog.Any("upstream == input == previousInput", inputW)))
-		return mtime.MinTimestamp, false, ptimeEventsReady, injectedReady
+
+	_, isOrdinaryStage := ss.kind.(*ordinaryStageKind)
+	if isOrdinaryStage && len(ss.sides) == 0 {
+		// For ordinary stage with no side inputs, we use whether there are pending elements to determine
+		// whether a bundle is ready or not.
+		if len(ss.pending) == 0 {
+			return mtime.MinTimestamp, false, ptimeEventsReady, injectedReady
+		}
+	} else {
+		if inputW == upstreamW && previousInputW == inputW {
+			slog.Debug("bundleReady: unchanged upstream watermark",
+				slog.String("stage", ss.ID),
+				slog.Group("watermark",
+					slog.Any("upstream == input == previousInput", inputW)))
+			return mtime.MinTimestamp, false, ptimeEventsReady, injectedReady
+		}
 	}
```

Comment:
We can avoid the extra indentation here without loss of meaning.

```suggestion
	} else if inputW == upstreamW && previousInputW == inputW {
		slog.Debug("bundleReady: unchanged upstream watermark",
			slog.String("stage", ss.ID),
			slog.Group("watermark",
				slog.Any("upstream == input == previousInput", inputW)))
		return mtime.MinTimestamp, false, ptimeEventsReady, injectedReady
}
```

Reply:
Done.